### PR TITLE
Support securing the Query frontend with OpenShift oauth-proxy

### PR DESCRIPTION
**Pull request by ccremer**
## Summary

* Adds a RBAC proxy sidecar to the Query frontend if enabled. This proxy does authentication (including redirect) via OpenShift OAuth and authorization via Kubernetes RBAC. This will restrict the exposed Query frontend to cluster administrators only.
* Parameter documentation

## Checklist

- [x] PR contains a single logical change (to build a better changelog).
- [x] Update the documentation.
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog.

### CODE REVIEW

File: component/query.libsonnet [hunk 1]
```diff
@@ -44,9 +47,101 @@ local query = thanos.query(queryBaseConfig + params.commonConfig + params.query
...
+        spec+: {
+          containers+: if params.queryRbacProxy.enabled then [
+            {
+              name: 'openshift-oauth-proxy',
+              image: '%s/%s:%s' % [ proxyImage.registry, proxyImage.image, proxyImage.tag ],
+              args: [
+                '--upstream=http://0.0.0.0:9090',
+                '--provider=openshift',
+                '--cookie-secret-file=/var/run/secrets/kubernetes.io/serviceaccount/token',
+                '--openshift-service-account=%s' % 'thanos-query',
```
[the remark this answers is not in the document]
Comment (simu) [reply component/query.libsonnet:73]: The following should work (IIRC how `super` works):

```suggestion
                '--openshift-service-account=%s' % super.spec.template.spec.serviceAccountName,
```

Otherwise you may have to do the following before the `spec+` on line 62:

```
local self = super;
```

and then use `self.spec.template.spec.serviceAccountName`.

File: component/query.libsonnet [hunk 2]
```diff
@@ -44,9 +47,69 @@ local query = thanos.query(queryBaseConfig + params.commonConfig + params.query
       type: params.query.serviceType,
     },
   },
+  serviceAccount+: {
+    metadata+: {
+      annotations+: if params.queryRbacProxy.enabled then {
+        'serviceaccounts.openshift.io/oauth-redirecturi.primary': params.queryRbacProxy.redirectUri,
```
Comment (simu) [inline component/query.libsonnet:53]: Again, this feels too OCP-specific to belong in a component which lives in organization projectsyn, any reason why we need openshift-specific annotations for the rbac proxy?

File: component/query.libsonnet [hunk 3]
```diff
@@ -44,9 +47,101 @@ local query = thanos.query(queryBaseConfig + params.commonConfig + params.query
...
+              args: [
+                '--upstream=http://0.0.0.0:9090',
+                '--provider=openshift',
+                '--cookie-secret-file=/var/run/secrets/kubernetes.io/serviceaccount/token',
+                '--openshift-service-account=%s' % 'thanos-query',
+                // TODO: Replace JSON below with `std.manifestJsonMinified({...}})` once available in newer jsonnet version
+                // Only cluster-admins should be able to get system resources
+                '--openshift-sar={"namespace":"%s","resource":"services","name":"%s","verb":"get"}' % [ params.namespace, params.queryRbacProxy.serviceName ],
+                '--http-address=http://:%s' % params.queryRbacProxy.port,
+                '--https-address=',
```
[the remark this answers is not in the document]
Comment (simu) [reply component/query.libsonnet:78]: I wouldn't spend the effort for re-encrypted traffic, since that would most likely be yet another OpenShift-specific feature in a component which should work on any K8s distribution

File: tests/defaults.yml [hunk 4]
```diff
@@ -1,7 +1,5 @@
 parameters:
   thanos:
-    commonConfig:
-      version: v0.23.0
```
Comment (simu) [inline tests/defaults.yml:4]: Why is this removed?

**Review by simu — CHANGES_REQUESTED** [review]
If we really need to use the OpenShift oauth-proxy (why is that in the first case?), then the `queryRbacProxy` feature must be gated behind a distribution check for OpenShift, e.g. something like `std.startsWith(parameters.facts.distribution, 'openshift')`, as the configuration provided in this PR certainly won't work on a distribution other than OpenShift 4.

File: class/defaults.yml [hunk 5]
```diff
@@ -62,3 +75,8 @@ parameters:
         registry: quay.io
         image: thanos/thanos
         tag: ${thanos:commonConfig:version}
+      oauthProxy:
+        # See supported tags with `oc -n openshift get imagestream oauth-proxy`
+        registry: image-registry.openshift-image-registry.svc:5000
+        image: openshift/oauth-proxy
+        tag: v4.4
```
[the remark this answers is not in the document]
Comment (ccremer) [reply class/defaults.yml:82]: I couldn't come up with a configuration where kube-rbac-proxy redirected the user to a login screen. It was just printing "Unauthorized" if the headers were missing.

I mean I could spend another day on figuring out a kube-native way...

File: tests/defaults.yml (hunk 4 above)
Comment (ccremer) [reply tests/defaults.yml:4]: I thought I'd use the default, which is still v0.18

File: component/query.libsonnet (hunk 2 above)
Comment (ccremer) [reply component/query.libsonnet:53]: This is needed to tell openshift-oauth where to redirect back to after authenticating.

I couldn't make the proxy k8s native... Either it's not possible with kube-rbac-proxy or I was screwing up the config

File: class/thanos.yml [hunk 6]
```diff
@@ -7,5 +7,6 @@ parameters:
         output_path: apps/
       - input_paths:
           - thanos/component/main.jsonnet
+          - thanos/component/ingress.jsonnet
```
[the remark this answers is not in the document]
Comment (ccremer) [reply class/thanos.yml:10]: no particular reason... I'm just used that way. I'll adjust

File: component/query.libsonnet (hunk 1 above)
[the remark this answers is not in the document]
Comment (ccremer) [reply component/query.libsonnet:73]: Your suggestion doesn't work:
```
Jsonnet error: failed to compile /tmp/component-m237v6r3/dependencies/thanos/component/main.jsonnet:
 RUNTIME ERROR: attempt to use super when there is no super class.
	/tmp/component-m237v6r3/dependencies/thanos/component/query.libsonnet:73:52-57	thunk <b>
```
And `local self = super;` `make format_jsonnet` yields
```
STATIC ERROR: component/query.libsonnet:62:11-15: expected token IDENTIFIER but got (self, "self")
```
```
  deployment+: {
    local self = super;
    spec+: {
```

[the remark this answers is not in the document]
Comment (simu) [reply component/query.libsonnet:73]: So, turns out I was mixing up two features, what works is the following:

In `deployment.spec.template.spec`:

```
local pod = self;
{
  containers+: [ ... ]
}
```

and then use `pod.spec.serviceAccountName` when configuring the sidecar.

[the remark this answers is not in the document]
Comment (ccremer) [reply component/query.libsonnet]: that worked, thanks!

File: class/defaults.yml [hunk 7]
```diff
@@ -38,6 +38,19 @@ parameters:
...
+      enabled: false
+      port: 8080
+      redirectUri: https://${thanos:ingress:host}
+    ingress:
+      enabled: false
+      serviceName: ${_instance}-auth
+      servicePort: ${thanos:queryRbacProxy:port}
+      annotations:
+        cert-manager.io/cluster-issuer: letsencrypt-production
+      secretName: ${_instance}-tls
```
[the remark this answers is not in the document]
Comment (ccremer) [reply class/defaults.yml:52]: Hm, fully exposing the Ingress object also means exposing arrays in TLS and backend configs. That would make it harder to configure hostnames via hierarchy.

I agree it's currently half-baked. Atm I'm opting for minimal config with only hostname, labels and annotations and make the `ingress` param a subkey of `queryRbacProxy` itself. WDYT?

[the remark this answers is not in the document]
Comment (simu) [reply class/defaults.yml:52]: I'm happy to have a minimal config in this particular case. Adding the parameter to `queryRbacProxy` is a good option to avoid the accessible-to-the world querier setup which is otherwise easily possible.

File: tests/golden/openshift-rbac-proxy/thanos/thanos/query/deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+            failureThreshold: 20
+            httpGet:
+              path: /-/ready
+              port: 9090
+              scheme: HTTP
+            periodSeconds: 5
+          resources: {}
+          terminationMessagePolicy: FallbackToLogsOnError
+        - args:
+            - --upstream=http://0.0.0.0:9090
```
Comment (bastjan) [inline tests/golden/openshift-rbac-proxy/thanos/thanos/query/deployment.yaml:79]: I'm really surprised this works. Shouldn't it be `127.0.0.1`?

Comment (ccremer) [reply tests/golden/openshift-rbac-proxy/thanos/thanos/query/deployment.yaml:79]: not really. The query container listens on 0.0.0.0 as you can inspect in the golden files:
```yaml
      containers:
        - args:
            - query
            - --grpc-address=0.0.0.0:10901
            - --http-address=0.0.0.0:9090
```

Comment (bastjan) [reply tests/golden/openshift-rbac-proxy/thanos/thanos/query/deployment.yaml:79]: I always thought `0.0.0.0` is not a valid destination address.

Apparently while technically not valid `0.0.0.0` may work in practice https://superuser.com/questions/949428/whats-the-difference-between-127-0-0-1-and-0-0-0-0 .

Comment (ccremer) [reply tests/golden/openshift-rbac-proxy/thanos/thanos/query/deployment.yaml:79]: Well, to quote your own article:

> In the context of servers, 0.0.0.0 means "all IPv4 addresses on the local machine"

If query would only bind loopback with `query - --http-address=127.0.0.1:9090` then it wouldn't be reachable from outside the container.

Comment (bastjan) [reply tests/golden/openshift-rbac-proxy/thanos/thanos/query/deployment.yaml:79]: Yes I'm not only talking about `--upstream=http://0.0.0.0:9090` since that's the client. The listener is configured with `--http-address=http://:8080` which tells go to listen on `0.0.0.0` on both IPv4 and IPv6.

**Review by bastjan — APPROVED** [review]

File: class/defaults.yml [hunk 9]
```diff
@@ -38,6 +38,15 @@ parameters:
       replicaLabels:
         - prometheus_replica
         - rule_replica
+    queryRbacProxy:
+      enabled: false
+      redirectUri: https://${thanos:queryRbacProxy:ingress:host}
```
Comment (simu) [inline class/defaults.yml:43]: I don't think `redirectUri` needs to be configurable, we can just use `queryRbacProxy.ingress.host` instead.

File: component/query.libsonnet [hunk 10]
```diff
@@ -44,9 +48,115 @@ local query = thanos.query(queryBaseConfig + params.commonConfig + params.query
       type: params.query.serviceType,
     },
   },
+  serviceAccount+: {
+    metadata+: {
+      annotations+: if params.queryRbacProxy.enabled then {
+        'serviceaccounts.openshift.io/oauth-redirecturi.primary': params.queryRbacProxy.redirectUri,
```
Comment (simu) [inline component/query.libsonnet:54]:
```suggestion
        'serviceaccounts.openshift.io/oauth-redirecturi.primary': 'https://' + params.queryRbacProxy.ingress.host,
```

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 11]
```diff
@@ -112,6 +112,70 @@ default:: `ClusterIP`
...
+
+Configuration for the RBAC proxy sidecar that authenticates requests to Thanos Query when exposed.
+Users are authenticated with OAuth and authorized using Kubernetes RBAC.
+It's recommended to deploy the proxy together with an Ingress.
+
+[NOTE]
+Only OpenShift is currently supported.
+
+The proxy redirects browser requests without authentication headers to OpenShift OAuth.
+Once authenticated, the proxy does authorization by issueing a `SubjectAccessReview` on behalf of the requesting user against the service named `queryRbacProxy.serviceName`.
```
Comment (simu) [inline docs/modules/ROOT/pages/references/parameters.adoc:128]:
```suggestion
Once authenticated, the proxy does authorization by issuing a `SubjectAccessReview` on behalf of the requesting user against the service named `queryRbacProxy.serviceName`.
```

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 12]
```diff
@@ -112,6 +112,70 @@ default:: `ClusterIP`
...
+
+=== `enabled`
+
+[horizontal]
+type:: bool
+default:: `false`
+
+Whether the RBAC should be deployed.
```
Comment (simu) [inline docs/modules/ROOT/pages/references/parameters.adoc:136]:
```suggestion
Whether the RBAC proxy should be deployed.
```

**Review by simu — APPROVED** [review]
LGTM overall
